2054 remove TSLint & upgrade to Angular9

benjamincharity opened the pull request:
- upgrade all deps
  - angular 9
  - typescript 3.8.3
  - rxjs 9
- get library built with ng commands
- ~get library tested with ng commands~
  - Currently not possible to pass flags to jest using ng.
- get library linted with ng commands
- remove unneeded tsconfig files
- update angular.json config
- remove tslint and all deps
- update to latest eslint package with new tslint replacement rules
- fix all lint items from the update
- stop shipping jest spies in prod code
- disable deprecation rule for some rxjs items for now (currently not sure they are valid deprecations)
- convert any instances of `it(` to `test(`
- move from circle to github workflows
- move bundled dependencies into peer dependencies (text-mask, scrollbar)
- update installation instructions (add new peer deps, add missing popper.js)


### STILL TODO (but no need to block review)

1. Write complete breaking migration notes
  - amcharts mock
  - any other jest mocks?
  - dependency update requirements (angular, ngrx, typescript)

---

BREAKING CHANGES

- Require Angular 9+
- Require TypeScript 3.8.3
- Require NGRX 9+
- No longer shipping jest spies inside mocks.
- Spec helpers that included jest code no longer exported
- Increased memory requirements during build

benjamincharity committed:
chore(Drawer): remove unused module.id

benjamincharity committed:
fix(Validators): no longer shipping jest spys

We still provide the same returns, but the consumer is responsible for spying.

BREAKING CHANGE:
ValidatorsService mock no longer defaults to a jest spy. Now it simply returns. It is up to the
consumer to spy if needed.

benjamincharity committed:
feat: Upgrade to Angular9 & remove TSLint

BREAKING CHANGE:

- TypeScript
- TSLint

benjamincharity committed:
refactor(LoginForm): update resetForm and increase test coverage

ISSUES CLOSED: #1287

atlwendy commented on projects/library/login-form/src/login-form.component.spec.ts:
```diff
@@ -1,202 +1,210 @@
-  });
 
-});
+// describe(`TsLoginFormComponent`, function() {
```
You planning to keep this commented out blocks?

benjamincharity replied:
Ahh nope.. this was a bad rebase 👍

atlwendy commented on projects/library/login-form/src/login-form.component.ts:
```diff
@@ -178,66 +134,58 @@ export class TsLoginFormComponent implements OnChanges {
+   */
+  public resetForm(): void {
+    this.loginForm.reset();
+    this.loginForm = this.createForm();
```
Shouldn't need this once the input pristine state is fixed 🤞

atlwendy commented on DEVELOPMENT.md:
```diff
@@ -75,17 +75,15 @@ $ yarn run contributors:generate
-1. `yarn install && yarn run build && yarn run start:app`
-2. Navigate to `http://localhost:4300/components/`
+1. `yarn install && yarn run library:build && yarn run demo:start`
+2. Navigate to `http://localhost:4200/components/`
```
Never realized the typo of `4300` here lol

atlwendy commented on angular.json:
```diff
@@ -23,7 +63,7 @@
             "polyfills": "projects/demo/src/polyfills.ts",
             "tsConfig": "projects/demo/tsconfig.app.json",
-            "aot": false,
+            "aot": true,
```
Does it mean that it'll always run with `aot`?

benjamincharity replied:
Yep

atlwendy commented on angular.json:
```diff
@@ -246,40 +289,13 @@
+    "@schematics/angular:component": {
+      "changeDetection": "OnPush",
+      "displayBlock": true,
+      "viewEncapsulation": "None"
```
This mean that if we use schematics to generate new component, these will be the default, right?

benjamincharity replied:
Yep

atlwendy commented on jest.base.config.js:
```diff
@@ -8,14 +8,14 @@ module.exports = {
   globals: {
     'ts-jest': {
-      tsConfig: './projects/library/tsconfig.spec.json',
+      tsConfig: 'projects/library/tsconfig.spec.json',
```
Do we now prefer not to have `./`? I remember we had this back and forth a couple times..

benjamincharity replied:
Yeah we never had a solid decision on it. Not sure why historically I've preferred the leading slash, but there's no solid reason for them and webstorm doesn't like them so I figured I accept it ;)

atlwendy commented on ngcc.config.js:
```diff
@@ -0,0 +1,13 @@
+      ignorableDeepImportMatchers: [
+        /text-mask-core\//,
+        /text-mask-addons\//,
+        /popper.js\//,
```
Interestingly these three are called out. When I dealt with ci, `text-mask-core`, '`text-mask-addons` and `ngx-perfect-scrollbar` were what it complained about.

benjamincharity replied:
These are the only three packages that we are actually shipping inside our code. We could always turn them into peer dependencies if it makes it easier. Doesn't really matter to the consumer past the install step. I originally just packaged them to make the install easier.

atlwendy commented on package.json:
```diff
@@ -13,60 +13,72 @@
-    "lint": "yarn run lint:eslint:fix && yarn run lint:tslint:fix && yarn run lint:scss:fix",
-    "lint:ci": "yarn run lint:eslint:ci && yarn run lint:tslint:ci && yarn run lint:scss",
+    "link": "cd dist/library && yarn link && cd ../../ && yarn link @terminus/ui",
+    "unlink": "yarn unlink @terminus/ui && cd dist/library && yarn unlink",
```
👍  Time to rebuild muscle memory lol

benjamincharity replied:
ohhh yeah lol

atlwendy commented on projects/demo/src/app/components/autocomplete/autocomplete-routing.module.ts:
```diff
@@ -1,5 +1,7 @@
-import { RouterModule, Routes } from '@angular/router';
+import {
+  RouterModule, Routes,
+} from '@angular/router';
```
Separate lines?

atlwendy commented on projects/demo/src/app/components/autocomplete/autocomplete.module.ts:
```diff
@@ -1,14 +1,16 @@
-import { FormsModule, ReactiveFormsModule } from '@angular/forms';
+import {
+  FormsModule, ReactiveFormsModule,
+} from '@angular/forms';
```
Separate lines?

atlwendy commented on projects/demo/src/app/components/autofocus/autofocus-routing.module.ts:
```diff
@@ -1,5 +1,7 @@
-import { RouterModule, Routes } from '@angular/router';
+import {
+  RouterModule, Routes,
+} from '@angular/router';
```
dang..
I think all the routing modules except button have the same line separating issues..

benjamincharity replied:
ahh dang.. yeah I didn't manually go through every demo file. Only the library files

benjamincharity replied:
Done

atlwendy commented on projects/demo/src/app/components/card/card.component.ts:
```diff
@@ -1,6 +1,9 @@
+} from '@angular/core';
+import {
+  TsCardBorderOptions, TsCardComponent,
+} from '@terminus/ui/card';
```
Line separating...

atlwendy commented on projects/demo/src/app/components/components.constant.ts:
```diff
@@ -4,296 +4,212 @@ import { Routes } from '@angular/router';
-    data: {
-      name: 'Autocomplete',
-    },
+    loadChildren: () => import('./autocomplete/autocomplete.module').then(m => m.AutocompleteModule),
```
Is this the v9 lazy loading style?

benjamincharity replied:
Yep

atlwendy commented on projects/library/autocomplete/src/autocomplete-panel/autocomplete-panel.component.ts:
```diff
@@ -86,8 +86,8 @@ export class TsAutocompletePanelComponent implements AfterContentInit {
    */
-  @ViewChild(TemplateRef, { static: false })
-  // tslint:disable-next-line no-any
+  @ViewChild(TemplateRef)
```
`static` object is not required any more?

benjamincharity replied:
In v9 static set to false by default so it only needs to be set when true

atlwendy commented on projects/library/autocomplete/src/autocomplete-panel/autocomplete-trigger.directive.ts:
```diff
@@ -229,13 +239,14 @@ export class TsAutocompleteTriggerDirective<ValueType = string> implements Contr
       this.closeKeyEventStream,
-      this.overlayRef.backdropClick(),
+      // eslint-disable-next-line deprecation/deprecation
+      this.overlayRef?.backdropClick() || of<string>(''),
```
Does this fall back happen when `this.overlayRef` is not defined?

benjamincharity replied:
In theory possibly, but in our actual codebase, there is no way to get to this point if the overlayRef doesn't exist so it can never be undefined.

atlwendy commented on projects/library/autocomplete/src/autocomplete-panel/autocomplete-trigger.directive.ts:
```diff
@@ -787,21 +804,26 @@ export class TsAutocompleteTriggerDirective<ValueType = string> implements Contr
     const optionChanges = this.autocompletePanel.options.changes.pipe(
+      // TODO: Refactor deprecation
+      // eslint-disable-next-line deprecation/deprecation
       tap(() => this.positionStrategy.reapplyLastPosition()),
```
even `tap` is deprecated?

benjamincharity replied:
Supposedly... need to dig more to really verify

atlwendy commented on projects/library/chip/src/chip-collection.component.ts:
```diff
@@ -167,32 +144,33 @@ export class TsChipCollectionComponent implements OnInit, AfterViewInit, AfterCo
    * @internal
    */
-  public selectionModel!: SelectionModel<TsChipComponent>;
+  public onChange: (value: string[]) => void = () => { };
```
What's the reason that these variables order rearranged?

benjamincharity replied:
Probably from trying to move the `member-order` rule over which requires properties before methods etc. I ended up having issues with it since it didn't like our private property to be with the getter/setter which we do in _every file_ so I had to pull the rule.

atlwendy commented on projects/library/chip/src/chip.module.ts:
```diff
@@ -2,13 +2,14 @@ import { CommonModule } from '@angular/common';
 
 export * from './chip.component';
 export * from './chip-collection.component';
+export * from './chip-badge.directive';
```
Uh oh... has the badge ever worked?

benjamincharity replied:
It was working in our demos... not sure how lol

atlwendy commented on projects/library/paginator/package.json:
```diff
@@ -4,7 +4,8 @@
         "@terminus/ngx-tools": "terminus.ngxTools",
-        "@terminus/ngx-tools/coercion": "terminus.ngxTools.coercion"
+        "@terminus/ngx-tools/coercion": "terminus.ngxTools.coercion",
+        "@terminus/ngx-tools/utilities": "terminus.ngxTools.utilities"
```
Wonder how it worked if `utilities` is needed but not included in the package?

benjamincharity replied:
Without it you'll see this warning during build:

![Screen Shot 2020-03-30 at 7 48 53 AM](https://user-images.githubusercontent.com/270193/77909204-f0280880-725a-11ea-9f3f-10a4a0b977b8.png)

And it usually guesses wrong since our modules are scoped. I haven't personally verified, but I'd expect this would affect consumers in certain scenarios.

atlwendy reviewed (commented):
looks good... just some lint/style of separating imports into multiple lines that you might want to update...

benjamincharity committed:
chore(CI): move to github workflow

chore(CI): don't run PR labeler action for forks
chore(CI): adjust sizes for auto PR labels
